src/uci_options.h

```diff
@@ -0,0 +1,159 @@
+#pragma once
+
+#include "parsing/input_parsing.h"
+
+#include "fmt/base.h"
+#include <cstdint>
+#include <functional>
+#include <optional>
+#include <string>
+#include <string_view>
+#include <variant>
+
+namespace ucioption {
+
+/* uci types for easier creation */
+using string = std::string;
+using check = bool;
+using spin = uint64_t;
+
+template<typename T>
+struct Storage {
+    T value;
+    const T defaultValue;
+    const std::function<void(const T&)> callback;
+};
+
+using Variant = std::variant<
+    Storage<std::string>,
+    Storage<bool>,
+    Storage<uint64_t>>;
+
+struct Limits {
+    uint64_t min;
+    uint64_t max;
+};
+
+struct UciOption {
+    const std::string_view name;
+    Variant variant;
+    const std::optional<Limits> limits {};
+};
+
+template<typename T>
+constexpr auto make(std::string_view name, T defaultVal, std::function<void(const T&)> cb)
+{
+    return UciOption {
+        .name = name,
+        .variant = Storage<T> { .value = defaultVal, .defaultValue = defaultVal, .callback = std::move(cb) },
+    };
+}
+
+template<typename T>
+constexpr auto make(std::string_view name, T defaultVal, Limits limits, std::function<void(const T&)> cb)
+{
+    return UciOption {
+        .name = name,
+        .variant = Storage<T> { .value = defaultVal, .defaultValue = defaultVal, .callback = std::move(cb) },
+        .limits = std::move(limits),
+    };
+}
+
+constexpr bool handleInput(UciOption& option, std::string_view input)
+{
+    return std::visit([input, name = option.name, limits = option.limits](auto&& arg) {
+        bool error = false;
+
+        using T = std::decay_t<decltype(arg)>;
+        if constexpr (std::is_same_v<T, Storage<bool>>) {
+            if (input == "true") {
+                arg.value = true;
+            } else if (input == "false") {
+                arg.value = false;
+            } else {
+                error = true;
+            }
+        } else if constexpr (std::is_same_v<T, Storage<std::string>>) {
+            arg.value = input;
+        } else if constexpr (std::is_same_v<T, Storage<uint64_t>>) {
+            const auto inputNum = parsing::to_number(input);
+            if (!inputNum.has_value())
+                return false;
+
+            if (limits.has_value()) {
+                if (*inputNum < limits->min || *inputNum > limits->max) {
+                    return false;
+                }
+            }
+
+            arg.value = *inputNum;
+        } else {
+            static_assert(false, "Uci option type handling has not been implemented yet!");
+        }
+
+        if (error) {
+            return error;
+        }
+
+        if (arg.callback) {
+            arg.callback(arg.value);
+        } else {
+            fmt::println("Callback is not set for {}", name);
```

Comment:
Should this be a static assert, in line with the above? This is a message for developers rather than users.

Reply:
The callback is not a constexpr, so it can not be evaluated at compile time.
I will change it to a runtime assert so we can handle it on debug level instead.

Just wanted a clear error here as it might be difficult to debug if something were to go wrong. Shouldn't happen though (famous last words..)